File: templates/matlab/likert_diverging.m
function fig = likert_diverging()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(2);
    questions = arrayfun(@(i)sprintf('Q%d',i), 1:6, 'UniformOutput', false);
    M = 5 + 20*rand(6, 5); M = M ./ sum(M, 2) * 100;
    cols = [palette('cat',2); palette('cat',8); 0.74*[1 1 1]; palette('cat',6); palette('cat',1)];
    fig = figure('Position',[100 100 800 400]); hold on;
    left = -(M(:, 1) + M(:, 2) + M(:, 3)/2);
    for k = 1:5
        for q = 1:6
            rectangle('Position', [left(q), q-0.4, M(q, k), 0.8], 'FaceColor', cols(k, :), 'EdgeColor','w');
        end
        left = left + M(:, k);
    end
    xline(0, 'k');
    set(gca,'YTick',1:6,'YTickLabel',questions);
    xlabel('% respondents'); title('Likert diverging bars');
end

File: templates/python/likert_diverging.py
"""likert_diverging: Likert 量表发散柱状（同意/反对中心对齐）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(title='Likert diverging bars'):
    apply_theme(fig_size=(8, 4))
    questions = [f'Q{i+1}' for i in range(6)]
    # strongly disagree, disagree, neutral, agree, strongly agree (%)
    rng = np.random.default_rng(2)
    M = rng.uniform(5, 25, (6, 5))
    M = M / M.sum(axis=1, keepdims=True) * 100
    colors = [cycle(1), cycle(7), '#bdbdbd', cycle(5), cycle(0)]
    fig, ax = plt.subplots()
    left = -(M[:, 0] + M[:, 1] + M[:, 2]/2)
    for k, c in enumerate(colors):
        ax.barh(questions, M[:, k], left=left, color=c, edgecolor='w', label=f'cat {k+1}')
        left += M[:, k]
    ax.axvline(0, color='k', linewidth=0.7)
    ax.set_xlabel('% of respondents'); ax.set_title(title)
    ax.legend(loc='upper center', bbox_to_anchor=(0.5, -0.12), ncol=5, fontsize=7)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)
